[VMware] Support cloudinit raw data feature

xiaofengw-vmware opened the pull request:
VMware product will allow user to specify their cloud-init meta data and user data to customize the VM.

 ## summary: 
This feature will modify VMware datasource to read from meta data and user data which are specified by VMware vSphere user. If meta data/user data are found in cloud-init configuration directory, datasource will parse the meta data/network and user data from the configuration file, otherwise it will continue to parse  them from traditional customization configuration file as before. The supported meta data file is in json or yaml format. 

## Test Steps
Unit tests are added to test_ovf.py and ensure get_data() works as expected.
For the end to end test, user need to deploy latest version of VMware vSphere and customize the VM with cloud-init raw meta data and user data. It will be covered by our integration test.

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [ x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

xiaofengw-vmware commented:
Please help to assign a reviewer and give us the comments, thanks a lot.

xiaofengw-vmware commented:
Can anybody help to review this pull request? Thanks a lot.

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+
+        userDataFile = cust_conf.user_data_name
+        if userDataFile:
+            userPath = util.log_time(
```
is there a reason to wait here? If metadatafile is present, can we just assume that it was written last? and thus userDataFile will already exist? 

It seems at least there is a race condition otherwise where we could read the metadataFile, and then not have a userDataFile for 10 seconds, but then one appear as soon as we stop looking.

It would be nice for there to be a guarantee from the platform in some way so we didn't have such a race.

Perhaps a third file "files.json" could list the files that were present, and then we could wait for that explicit list? or some other way so we don't potentially go on without finding a file that would have been there later.

xiaofengw-vmware replied:
Thanks, I will add these files as a list and call util.wait_for_files(). Also, we will ensure meta data is not last file to be copied.

xiaofengw-vmware replied:
The vmware platform will ensure cust.cfg is copied as the last file， so we could assume these 3 files are copied there, here we needn't to wait for any files. I will change the codes accordingly.

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -378,15 +434,16 @@ def get_max_wait_from_cfg(cfg):
+def wait_for_imc_file(filename, maxwait=180, naplen=5,
+                      dirpath="/var/run/vmware-imc"):
     waited = 0
 
```
there is a generic 'wait_for_files' in utils that you could use.

```python

missing = util.wait_for_files(
    [os.path.join(dirpath, filename)], maxwait,
    naplen=naplen, "VMware customization: ")
return len(missing) == 0
```

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    ud = None
+    network = None
+
+    md = load(util.load_file(metaPath).replace("\r", ""))
```
it seems wrong to do the 'replace' of \r here.  Either the content of that file was valid json/yaml or it wasnt.  A simple 'replace' is not going to know the difference between a \r in yaml and a \r as "windows end of line".

xiaofengw-vmware replied:
fixed.

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    pass
+
+
+class FileNotFound(Exception):
```
just drop this and use FileNotFoundError (we don't support python2 any more).

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    return md, ud, cfg, network
+
+
+class LoadError(Exception):
```
If you change the loader to just load with yaml or json, then you can drop this.  It doesn't actually provide us with any additional information , in fact your use of it *loses* information as the original error had info on what was invalid (json parse or yaml parse).

xiaofengw-vmware replied:
Fixed

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -76,7 +78,7 @@ def _get_data(self):
         vd = ""
         vmwareImcConfigFilePath = None
-        nicspath = None
+        nicsPath = None
```
generally python is snake_case, not camelCase, i don't think the change here actually helps anything.

xiaofengw-vmware replied:
Revert it back, thanks.

smoser reviewed (dismissed):
I'm really happy to see this, and I hope another reviewer (@OddBloke or @blackboxsw ) have time to look at this.  Having cloud-init support in Vsphere is really a huge win.

I've done a cursory look.  There are some changes needed.

raharper commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -124,31 +124,87 @@ def _get_data(self):
-                LOG.debug("Customization for VMware platform is disabled.")
 
-        if vmwareImcConfigFilePath:
+                # if meta data is avaiable, ignore disable_vmware_customization
```
```suggestion
                # if meta data is available ignore disable_vmware_customization
```

xiaofengw-vmware replied:
> 

Fixed.

raharper commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+
+    if 'network' in md:
+        network = md['network']
+        del md['network']
```
Why remove the key from the metadata dictionary ?

xiaofengw-vmware replied:
The network in md is saved into network, so we still need to keep it in md?

blackboxsw replied:
@xiaofengw-vmware. cloud-init caches metadata is reads from the datasource and stores it in /var/lib/cloud/instance/obj.pkl as the cached datasource. the utility `cloud-init query ds` returns this datasource metadata JSON info to customers (and developers) who might need to debug the metadata content provided by the underlying datasource to see what is affecting the config on an instance. It is helpful to preserve as close to raw content as possible for better bug triage and analysis on the cloud-init deployed machines. If possible, please leave this key in the metadata in general. Otherwise on debug folks need to load the obj,pkl and inspect _network_config on the cached datasource instead of just `cloud-init query ds.meta_data.network`

xiaofengw-vmware commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +741,98 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    '''
+    first attempts to unmarshal the provided data as JSON, and if
+    that fails then attempts to unmarshal the data as YAML. If data is
+    None then a new dictionary is returned.
```
Thanks Scott, I will choose a). because we want to support both json and yaml format.

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,73 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    else:
+        nicsPath = os.path.join(VMWARE_IMC_DIR, "nics.txt")
+        if not os.path.exists(nicsPath):
+            LOG.debug('nics.txt is not exist.')
```
Probably worth actually parameterizing these logs since your log messages redact the full path of metadata, userdat and nics.txt.
```suggestion
            LOG.debug('%s does not exist.', nics_path)
```

xiaofengw-vmware replied:
Fixed, PTAL.

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,73 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+        if userDataFile:
+            userPath = os.path.join(VMWARE_IMC_DIR, userDataFile)
+            if not os.path.exists(userPath):
+                raise FileNotFoundError("user data file is not found")
```
it's best to parameterize the log message to give more specific full paths in context
```suggestion
                raise FileNotFoundError("user data file is not found %s", ud_path)
```

xiaofengw-vmware replied:
Fixed, PTAL.

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -129,26 +129,82 @@ def _get_data(self):
+                        vmwareImcConfigFilePath = None
+
+        if vmwareImcConfigFilePath and metaPath:
+            try:
+                set_gc_status(self._vmware_cust_conf, "Started")
+
+                LOG.debug("Start to load cloud-init meta data and user data")
+                (md, ud, cfg, network) = load_cloudinit_data(
+                    metaPath, userPath)
+
+                if network:
+                    self._network_config = network
+                else:
+                    fallbackNetwork = self.distro.generate_fallback_config()
+                    self._network_config = fallbackNetwork
+
+            except safeyaml.YAMLError as e:
```
A couple of minor points here:
 - generally it's nice to only include operations within the try/except blocks that are likely to generate the exceptions we are handling. So we should move things like LOG.deb and set_gc_status outside of the try. It makes it easier to understand what callsites are likely to raise errors
 -  I don't think we need to create and assign a fallbackNetwork local variable as we aren't using that anywhere else

The changeset I was thinking is here

```suggestion
            LOG.debug("Start to load cloud-init meta data and user data")
            set_gc_status(self._vmware_cust_conf, "Started")
            try:
                (md, ud, cfg, network) = load_cloudinit_data(md_path, ud_path)
                if network:
                    self._network_config = network
                else:
                    self._network_config = (
                        self.distro.generate_fallback_config()
                    )
            except safeyaml.YAMLError as e:
```

xiaofengw-vmware replied:
Fixed. Your points 2) is quite useful. I created a local var because I couldn't write "self._network_config = self.distro.generate_fallback_config" in one line. From your comments, I learnt that we could use () to split it into 2 lines. Thanks a lot.

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,73 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    ud = None
+    network = None
+
+    md = load(util.load_file(metaPath))
```
I think we can drop both "def load and this callsite and just use util.read_conf(md_path) which will default to {} if file doesn't exist, is unparseable as YAML (emitting a more specific YAML format warnings in logs).

xiaofengw-vmware replied:
great and I fixed it, PTAL.

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,73 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    return md, ud, cfg, network
+
+
+def load(data):
```
might be able to drop in favor of util.read_conf which behaves with this default {} for unparseable or empty files.

xiaofengw-vmware replied:
util.read_conf() doesn't raise exception if it's not a valid yaml, it will log a warning trace and return {}. But here we want to catch the format exception and ask our vSphere user to modify it.

blackboxsw replied:
+1 let's rename this function to give a bit more clarity:
```suggestion
def safeload_yaml_or_dict(data):
```

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,73 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+def collect_imc_files(cust_conf):
+    '''
+    collect all the other imc files. Since these files are copied
+    before cust.cfg is copied, check if they exist or not directly.
```
I'm not quite sure about the doctstring meaning with "not directly"?

Also from the code it looks like metadata/userdata are mutually exclusive from nics.txt. If this is true, let's qualify the docstring to say either metadata/userdata or nics.txt must be present.

xiaofengw-vmware replied:
metadata/userdata must be present if they are specified in customization configuration. In this case, nics.txt is not needed. For other case, we should try to load nics.txt (optional)

blackboxsw reviewed (changes requested):
Thank you for taking this on @xiaofengw-vmware I've given your branch a review too with a couple of comments as well to try to cut down on some duplication of cloudinit.util functions if we can. Again this looks like a good approach for cloud-init users let's polish it a bit and get it into master :)

xiaofengw-vmware commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,73 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    - nics.txt:
+      optional if customization specification is traditional one
+    '''
+    metaPath = None
+    userPath = None
+    nicsPath = None
+    metaDataFile = cust_conf.meta_data_name
+    if metaDataFile:
```
Fixed, please take another look. Thanks.

xiaofengw-vmware commented:
> Thank you for taking this on @xiaofengw-vmware I've given your branch a review too with a couple of comments as well to try to cut down on some duplication of cloudinit.util functions if we can. Again this looks like a good approach for cloud-init users let's polish it a bit and get it into master :)

Thanks Chad for your quick response, I just addressed your comments, please have another review. Thanks for your time.

xiaofengw-vmware commented:
Hi all,
Sorry to disturb you. I know it's close to the holiday, I wonder whether you could give me another review before the holiday so that I could deliver the codes within this month. Thank you all for the help.

smoser commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -99,9 +101,7 @@ def _get_data(self):
                 LOG.debug("Skipping the check for "
                           "VMware Customization support")
-            elif not util.get_cfg_option_bool(
-                    self.sys_cfg, "disable_vmware_customization", True):
```
It appears that you have dropped the ability to disable vmware customization. This should be considered very carefully, as the previous value was `True`.  I don't have resources (time or vmware) to fully consider this, but it feels like this is a big change.

You'll need to comment on why that is OK, and adjust tests that reference it (tests/unittests/test_datasource/test_ovf).  If you're dropping the configuration setting entirely, then you'd want to remove all references to it across the code base.

xiaofengw-vmware commented:
Hi Scott,
Sorry I didn't click the "resolve conversation" button so you couldn't see my reply.  Please help to check the reply when you have time. Thanks a lot.

OddBloke commented:
Sounds like this is waiting for another pass from Scott and/or Chad?

xiaofengw-vmware commented:
> Sounds like this is waiting for another pass from Scott and/or Chad?

Yes, I modify the codes per their comments and need their approval.

xiaofengw-vmware commented:
Hi all,
Happy new year! Please help to take another look when you are back from the holiday. Thanks a lot. :)

smoser commented on tests/unittests/test_datasource/test_ovf.py:
```diff
@@ -138,16 +139,29 @@ def test_get_data_false_on_none_dmi_data(self):
 
     def test_get_data_no_vmware_customization_disabled(self):
-        """When vmware customization is disabled via sys_cfg log a message."""
+        """When cloud-init workflow for vmware is disabled via sys_cfgi and
```
typo (sys_cfgi)

xiaofengw-vmware replied:
Thanks, this is fixed and please take another look.

xiaofengw-vmware commented:
Hi all,
The changes are made per Scott's comments, PTAL.
Thanks.

xiaofengw-vmware commented:
Hi all,
Sorry to disturb you again. Could you help to review  the changes?  We want to deliver the codes as early as possible so that we could start the integration test. Thanks for your time. :)
@smoser @raharper @blackboxsw

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -129,26 +129,82 @@ def _get_data(self):
-                    nicspath = wait_for_imc_cfg_file(
-                        filename="nics.txt", maxwait=10, naplen=5)
+                    try:
+                        (md_path, ud_path, nicspath) = collect_imc_files(
```
let's rename this to collect_imc_file_paths so it's clearer what this function returns (returns paths instead of file contents)
```suggestion
                        (md_path, ud_path, nicspath) = collect_imc_file_paths(
```

xiaofengw-vmware replied:
Thanks Chad for the detail comments, I addressed all of them. PTAL.

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,69 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    return safeyaml.load(data)
+
+
+def collect_imc_files(cust_conf):
+    '''
+    collect all the other imc files. metadata/userdata must be present
+    if they are specified in customization configuration.
+    '''
```
let's rename this to collect_imc_file_paths so it's clearer what this function returns (returns paths instead of file contents)
Also let's add a bit more to the docstring so "future us" can understand the intent here

```suggestion
def collect_imc_file_paths(cust_conf):
    '''
    collect all the other imc files
    
    metadata is preferred to nics.txt configuration data.

    If metadata file exists because it is specified in customization
    configuration, then both metadata an userdata are required.

    @return a 3-tuple containing desired configuration file paths if present
        Expected returns:
             1. user provided metadata (md_path, ud_path, None)
             2. user-provided network config (None, None, nics_path)
             3. No config found (None, None, None)
    '''
```

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,69 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    """
+    Load the cloud-init meta data, user data, cfg and network from the
+    given files
+    """
```
more helpful docstrings
```suggestion

    @return: 4-tuple of configuration
        metadata, userdata, cfg={}, network
    @raises: FileNotFoundError if md_path or ud_path are absent
    """
```

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,69 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    LOG.debug('load meta data from: %s: user data from: %s',
+              md_path, ud_path)
+    md = {}
+    cfg = {}
```
Don't need this local var, just return {} as the 3rd element in the return tuple

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,69 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+
+    if 'network' in md:
+        network = md['network']
+        del md['network']
```
```suggestion
```

blackboxsw commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -684,4 +740,69 @@ def _raise_error_status(prefix, error, event, config_file, conf):
+    else:
+        nics_path = os.path.join(VMWARE_IMC_DIR, "nics.txt")
+        if not os.path.exists(nics_path):
+            LOG.debug('%s is not exist.', nics_path)
```
minor typo
```suggestion
            LOG.debug('%s does not exist.', nics_path)
```

blackboxsw reviewed (changes requested):
@xiaofengw-vmware Again thanks for your patience here. I got a bit wrapped up with holidays on this.

I've made individual comments on your PR in this last round to try to improve a bit of maintainability of the code/docs for when we look at this in the future and forget the intent of some of this logic.

Here is the consolidated diff of my comments for you to review and consider what you think is applicable.
```diff
diff --git a/cloudinit/sources/DataSourceOVF.py b/cloudinit/sources/DataSourceOVF.py
index da8d832d..d3a50113 100644
--- a/cloudinit/sources/DataSourceOVF.py
+++ b/cloudinit/sources/DataSourceOVF.py
@@ -119,7 +119,9 @@ class DataSourceOVF(sources.DataSource):
                     # When the VM is powered on, the "VMware Tools" daemon
                     # copies the customization specification file to
                     # /var/run/vmware-imc directory. cloud-init code needs
-                    # to search for the file in that directory.
+                    # to search for the file in that directory which indicates
+                    # that required metadata and userdata files are now
+                    # present.
                     max_wait = get_max_wait_from_cfg(self.ds_cfg)
                     vmwareImcConfigFilePath = util.log_time(
                         logfunc=LOG.debug,
@@ -137,9 +139,9 @@ class DataSourceOVF(sources.DataSource):
                     LOG.debug("Found VMware Customization Config File at %s",
                               vmwareImcConfigFilePath)
                     try:
-                        (md_path, ud_path, nicspath) = collect_imc_files(
+                        (md_path, ud_path, nicspath) = collect_imc_file_paths(
                             self._vmware_cust_conf)
-                    except Exception as e:
+                    except FileNotFoundError as e:
                         _raise_error_status(
                             "File(s) missing in directory",
                             e,
@@ -149,7 +151,7 @@ class DataSourceOVF(sources.DataSource):
                 else:
                     LOG.debug("Did not find VMware Customization Config File")
 
-                # ignore disable_vmware_customization if meta data is available
+                # Honor disable_vmware_customization setting on metadata absent
                 if not md_path:
                     if util.get_cfg_option_bool(self.sys_cfg,
                                                 "disable_vmware_customization",
@@ -159,13 +161,12 @@ class DataSourceOVF(sources.DataSource):
                         # reset vmwareImcConfigFilePath to None to avoid
                         # customization for VMware platform
                         vmwareImcConfigFilePath = None
-
-        if vmwareImcConfigFilePath and md_path:
+        use_raw_userdata = bool(vmwareImcConfigFilePath and md_path)
+        if use_raw_userdata:
             set_gc_status(self._vmware_cust_conf, "Started")
             LOG.debug("Start to load cloud-init meta data and user data")
             try:
-                (md, ud, cfg, network) = load_cloudinit_data(
-                    md_path, ud_path)
+                (md, ud, cfg, network) = load_cloudinit_data(md_path, ud_path)
 
                 if network:
                     self._network_config = network
@@ -199,6 +200,7 @@ class DataSourceOVF(sources.DataSource):
             set_gc_status(self._vmware_cust_conf, "Successful")
 
         elif vmwareImcConfigFilePath:
+            # Load configuration from vmware_imc
             self._vmware_nics_to_enable = ""
             try:
                 set_gc_status(self._vmware_cust_conf, "Started")
@@ -745,26 +747,29 @@ def load_cloudinit_data(md_path, ud_path):
     """
     Load the cloud-init meta data, user data, cfg and network from the
     given files
+
+    @return: 4-tuple of configuration
+        metadata, userdata, cfg={}, network
+
+    @raises: FileNotFoundError if md_path or ud_path are absent
     """
     LOG.debug('load meta data from: %s: user data from: %s',
               md_path, ud_path)
     md = {}
-    cfg = {}
     ud = None
     network = None
 
-    md = load(util.load_file(md_path))
+    md = safeload_yaml_or_dict(util.load_file(md_path))
 
     if 'network' in md:
         network = md['network']
-        del md['network']
 
     if ud_path:
         ud = util.load_file(ud_path).replace("\r", "")
-    return md, ud, cfg, network
+    return md, ud, {}, network
 
 
-def load(data):
+def safeload_yaml_or_dict(data):
     '''
     The meta data could be JSON or YAML. Since YAML is a strict superset of
     JSON, we will unmarshal the data as YAML. If data is None then a new
@@ -775,10 +780,20 @@ def load(data):
     return safeyaml.load(data)
 
 
-def collect_imc_files(cust_conf):
+def collect_imc_file_paths(cust_conf):
     '''
-    collect all the other imc files. metadata/userdata must be present
-    if they are specified in customization configuration.
+    collect all the other imc files.
+
+    metadata is preferred to nics.txt configuration data.
+
+    If metadata file exists because it is specified in customization
+    configuration, then both metadata an userdata are required.
+
+    @return a 3-tuple containing desired configuration file paths if present
+        Expected returns:
+             1. user provided metadata (md_path, ud_path, None)
+             2. user-provided network config (None, None, nics_path)
+             3. No config found (None, None, None)
     '''
     md_path = None
     ud_path = None
@@ -799,7 +814,7 @@ def collect_imc_files(cust_conf):
     else:
         nics_path = os.path.join(VMWARE_IMC_DIR, "nics.txt")
         if not os.path.exists(nics_path):
-            LOG.debug('%s is not exist.', nics_path)
+            LOG.debug('%s does not exist.', nics_path)
             nics_path = None
 
     return md_path, ud_path, nics_path
```

ubuntu-server-builder commented on cloudinit/sources/DataSourceOVF.py:
```diff
@@ -129,26 +129,82 @@ def _get_data(self):
+                        # customization for VMware platform
+                        vmwareImcConfigFilePath = None
+
+        if vmwareImcConfigFilePath and md_path:
```
Can we also name this condition something more understandable about the intent?
```suggestion
        use_raw_userdata = bool(vmwareImcConfigFilePath and md_path)
        if use_raw_userdata:
```

xiaofengw-vmware replied:
Here I change use_raw_userdata  to use_raw_data because userdata is part of the raw data. Here I want to refer to raw_data.

blackboxsw reviewed (approved):
Thank you so much for working to get this in @xiaofengw-vmware . Let's land it!

mweinelt commented:
@xiaofengw-vmware Can you clarify what version of VSphere is required to use raw cloud-init data?

xiaofengw-vmware commented:
> @xiaofengw-vmware Can you clarify what version of VSphere is required to use raw cloud-init data?

It will be supported from vSphere 7.0 update 3 which will be released soon. Thanks.
